README.md
# baseconv - Base conversion

[![Base conversion](https://raw.githubusercontent.com/godsfood/baseconv/master/.img/baseconv.png)](https://github.com/godsfood/baseconv)

[![Version](https://img.shields.io/github/v/tag/godsfood/baseconv)](https://pkg.go.dev/go.dkinom.dev/baseconv)
[![Build Status](https://github.com/godsfood/baseconv/actions/workflows/.github/workflows/go.yml/badge.svg?branch=master)](https://github.com/godsfood/baseconv/actions)
[![Coverage Status](https://coveralls.io/repos/github/godsfood/baseconv/badge.svg)](https://coveralls.io/github/godsfood/baseconv)
[![License](https://img.shields.io/badge/license-MIT-green)](https://github.com/godsfood/baseconv/blob/master/LICENSE)

A Go module for converting between different bases, e.g., decimal ↔ binary, octal ↔ hexadecimal

## Installing

```bash
go get go.dkinom.dev/baseconv
```

## Usage

See `example/main.go`

#### Hexadecimal ↔ Base58
```go
hexadecimalToBase58Converter, _ := baseconv.NewBaseConversion("0123456789abcdef", baseconv.Base58)

base58Value, _ := hexadecimalToBase58Converter.Convert("415a59758fb933b6049b050a556dd4d916b7b483f6966615")
// base58Value == "6xZA4Qt9vH7rePWeT5WLaVUZNjB6u6rGc"

hexadecimal, _ := hexadecimalToBase58Converter.Inverse().Convert("GjWGF6jERR9ymrC1bHcGmsJYkLMDoaySr")
// hexadecimal == "ac93c8d619c76f823f184110759b278f246cc7cc3cadcac3"
```

#### Decimal (emoji) ↔ Hexadecimal
```go
decimalEmojiToHexadecimalConverter, _ := baseconv.NewBaseConversion(
  "0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣", baseconv.Base16,
  options.BaseConversion().
    SetZeroPadding(true),
)

hexadecimal, _ := decimalEmojiToHexadecimalConverter.Convert("5️⃣1️⃣9️⃣6️⃣6️⃣")
// hexadecimal == "0CAFE"

decimalEmoji, _ := decimalEmojiToHexadecimalConverter.Inverse().Convert("DEADC0DE")
// decimalEmoji == "3️⃣7️⃣3️⃣5️⃣9️⃣2️⃣9️⃣0️⃣5️⃣4️⃣"
```

### Exported alphabets
- `Base2` - `01`
- `Base3` - `012`
- `Base4` - `0123`
- `Base5` - `01234`
- `Base6` - `012345`
- `Base8` - `01234567`
- `Base10` - `0123456789`
- `Base12` - `0123456789AB`
- `Base16` - `0123456789ABCDEF`
- `Base32` - `ABCDEFGHIJKLMNOPQRSTUVWXYZ234567`
- `Base32hex` - `0123456789ABCDEFGHIJKLMNOPQRSTUV`
- `Base36` - `0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ`
- `Base58` - `123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz`
- `Base62` - `0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz`
- `Base64` - `ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/`
- `Base64url` - `ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_`

## Syntax

### *`NewBaseConversion(from string, to string, opts ...*options.BaseConversionOptions) (*baseConversion, error)`*

*`from`* - String of numeral symbols representing the digits of `from` numeral system.

*`to`* - String of numeral symbols representing the digits of `to` numeral system.

See [documentation](https://pkg.go.dev/go.dkinom.dev/baseconv#section-documentation) for more

## License

[MIT](https://github.com/godsfood/baseconv/blob/master/LICENSE)


CHANGELOG.md
## 0.2.4

- Deleted `go.dkinom.dev/baseconv/characters`
- Added tests in `go.dkinom.dev/baseconv/internal/characters`

## 0.2.3

- Retracted `go.dkinom.dev/baseconv/characters`
- Added tests in `go.dkinom.dev/baseconv/options`

## 0.2.2

- Optimization for ASCII only alphabets
- Added benchmarks

## 0.2.1

- Added zero padding option

## 0.2.0

- Vanity import `go.dkinom.dev/baseconv`

## 0.1.0

- Initial development release

## 0.1.0-rc.0

- Release candidate for `0.1.0`


alphabet_test.go
package baseconv

import (
	"testing"
)

func TestAlphabetValidations(t *testing.T) {
	t.Run("alphabet has duplicate characters", func(t *testing.T) {
		alphabet, err := NewAlphabet("0123453789")

		if alphabet != nil || err == nil {
			t.Fatalf("alphabet '%v' has duplicate characters", alphabet)
		}
	})
}

func TestAlphabetString(t *testing.T) {
	t.Run("base-10", func(t *testing.T) {
		alphabet, err := NewAlphabet("0123456789")
		if err != nil {
			t.Fatal(err)
		}
		expected := "0,1,2,3,4,5,6,7,8,9"
		if alphabet.String() != expected {
			t.Fatal()
		}
	})
}

func TestAlphabetRadix(t *testing.T) {
	t.Run("base-10", func(t *testing.T) {
		alphabet, err := NewAlphabet("0123456789")
		if err != nil {
			t.Fatal(err)
		}
		expected := 10
		if alphabet.Radix() != expected {
			t.Fatalf("radix of '%v' is %v", alphabet, expected)
		}
	})

	t.Run("base-2", func(t *testing.T) {
		alphabet, err := NewAlphabet("01")
		if err != nil {
			t.Fatal(err)
		}
		expected := 2
		if alphabet.Radix() != expected {
			t.Fatalf("radix of '%v' is %v", alphabet, expected)
		}
	})

	t.Run("base-16", func(t *testing.T) {
		alphabet, err := NewAlphabet("0123456789abcdef")
		if err != nil {
			t.Fatal(err)
		}
		expected := 16
		if alphabet.Radix() != expected {
			t.Fatalf("radix of '%v' is %v", alphabet, expected)
		}
	})

	t.Run("base-10 (emoji)", func(t *testing.T) {
		alphabet, err := NewAlphabet("0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣")
		if err != nil {
			t.Fatal(err)
		}
		expected := 10
		if alphabet.Radix() != expected {
			t.Fatalf("radix of '%v' is %v", alphabet, expected)
		}
	})
}

func TestAlphabetIsValid(t *testing.T) {
	t.Run("valid base-10 numerals", func(t *testing.T) {
		alphabet, err := NewAlphabet("0123456789")
		if err != nil {
			t.Fatal(err)
		}

		validNumerals := []string{
			"123",
		}

		for _, n := range validNumerals {
			if !alphabet.IsValid(n) {
				t.Fatalf("'%v' is valid", n)
			}
		}
	})

	t.Run("invalid base-10 numerals", func(t *testing.T) {
		alphabet, err := NewAlphabet("0123456789")
		if err != nil {
			t.Fatal(err)
		}

		invalidNumerals := []string{
			"",
			"123abc",
		}

		for _, n := range invalidNumerals {
			if alphabet.IsValid(n) {
				t.Fatalf("'%v' is invalid", n)
			}
		}
	})

	t.Run("valid base-10 (emoji) numerals", func(t *testing.T) {
		alphabet, err := NewAlphabet("0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣")
		if err != nil {
			t.Fatal(err)
		}

		validNumerals := []string{
			"2️⃣6️⃣4️⃣8️⃣",
		}

		for _, n := range validNumerals {
			if !alphabet.IsValid(n) {
				t.Fatalf("'%v' is valid", n)
			}
		}
	})

	t.Run("invalid base-10 (emoji) numerals", func(t *testing.T) {
		alphabet, err := NewAlphabet("0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣")
		if err != nil {
			t.Fatal(err)
		}

		invalidNumerals := []string{
			"",
			"2️⃣6️⃣4️⃣8️⃣🔢",
		}

		for _, n := range invalidNumerals {
			if alphabet.IsValid(n) {
				t.Fatalf("'%v' is invalid", n)
			}
		}
	})
}


base_conversion_test.go
package baseconv

import (
	"fmt"
	"testing"

	"go.dkinom.dev/baseconv/options"
)

func TestValidations(t *testing.T) {
	t.Run("alphabet has duplicate characters 1", func(t *testing.T) {
		alphabet := "0123453789"
		b, err := NewBaseConversion(Hexadecimal, alphabet)

		if b != nil || err == nil {
			t.Fatalf("alphabet '%v' has duplicate characters", alphabet)
		}
	})

	t.Run("alphabet has duplicate characters 2", func(t *testing.T) {
		alphabet := "0123453789"
		b, err := NewBaseConversion(alphabet, Binary)

		if b != nil || err == nil {
			t.Fatalf("alphabet '%v' has duplicate characters", alphabet)
		}
	})
}

func TestConverter(t *testing.T) {
	t.Run("invalid input", func(t *testing.T) {
		b, err := NewBaseConversion(Decimal, Binary)
		if err != nil {
			t.Fatal(err)
		}

		s := "123abc"
		r, err := b.Convert(s)

		if r != "" || err == nil {
			t.Fatalf("'%v' is invalid", s)
		}
	})

	{
		b, err := NewBaseConversion(Decimal, Binary)
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b,
			"testCases": []map[string]string{
				{"from": "0", "to": "0"},
				{"from": "1", "to": "1"},
				{"from": "2", "to": "10"},
				{"from": "3", "to": "11"},
				{"from": "4", "to": "100"},
				{"from": "5", "to": "101"},
				{"from": "6", "to": "110"},
				{"from": "7", "to": "111"},
				{"from": "8", "to": "1000"},
				{"from": "9", "to": "1001"},
				{"from": "10", "to": "1010"},
			},
		}
		testConverter(t, converter)
	}
	{
		b, err := NewBaseConversion(
			Decimal, Binary,
			nil, options.BaseConversion().SetZeroPadding(true), nil,
		)
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b,
			"testCases": []map[string]string{
				{"from": "0", "to": "0000"},
				{"from": "1", "to": "0001"},
				{"from": "2", "to": "0010"},
				{"from": "3", "to": "0011"},
				{"from": "4", "to": "0100"},
				{"from": "5", "to": "0101"},
				{"from": "6", "to": "0110"},
				{"from": "7", "to": "0111"},
				{"from": "8", "to": "1000"},
				{"from": "9", "to": "1001"},
				{"from": "10", "to": "0001010"},
			},
		}
		testConverter(t, converter)
	}
	{
		b, err := NewBaseConversion(
			Decimal, Binary,
			options.BaseConversion().SetZeroPadding(true),
		)
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b.Inverse(),
			"testCases": []map[string]string{
				{"to": "00", "from": "0000"},
				{"to": "01", "from": "0001"},
				{"to": "02", "from": "0010"},
				{"to": "03", "from": "0011"},
				{"to": "04", "from": "0100"},
				{"to": "05", "from": "0101"},
				{"to": "06", "from": "0110"},
				{"to": "07", "from": "0111"},
				{"to": "08", "from": "1000"},
				{"to": "09", "from": "1001"},
				{"to": "010", "from": "0001010"},
			},
		}
		testConverter(t, converter)
	}
	{
		b, err := NewBaseConversion(Decimal, "0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣")
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b,
			"testCases": []map[string]string{
				{"from": "0", "to": "0️⃣"},
				{"from": "1", "to": "1️⃣"},
				{"from": "2", "to": "2️⃣"},
				{"from": "3", "to": "3️⃣"},
				{"from": "4", "to": "4️⃣"},
				{"from": "5", "to": "5️⃣"},
				{"from": "6", "to": "6️⃣"},
				{"from": "7", "to": "7️⃣"},
				{"from": "8", "to": "8️⃣"},
				{"from": "9", "to": "9️⃣"},
				{"from": "10", "to": "1️⃣0️⃣"},
			},
		}
		testConverter(t, converter)
	}
	{
		b, err := NewBaseConversion(
			Decimal, "0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣",
			nil, options.BaseConversion().SetZeroPadding(true), nil,
		)
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b,
			"testCases": []map[string]string{
				{"from": "0", "to": "0️⃣"},
				{"from": "1", "to": "1️⃣"},
				{"from": "2", "to": "2️⃣"},
				{"from": "3", "to": "3️⃣"},
				{"from": "4", "to": "4️⃣"},
				{"from": "5", "to": "5️⃣"},
				{"from": "6", "to": "6️⃣"},
				{"from": "7", "to": "7️⃣"},
				{"from": "8", "to": "8️⃣"},
				{"from": "9", "to": "9️⃣"},
				{"from": "10", "to": "1️⃣0️⃣"},
			},
		}
		testConverter(t, converter)
	}
	{
		b, err := NewBaseConversion(
			Hexadecimal, "0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣",
			options.BaseConversion().SetZeroPadding(true),
		)
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b.Inverse(),
			"testCases": []map[string]string{
				{"to": "0CAFE", "from": "5️⃣1️⃣9️⃣6️⃣6️⃣"},
				{"to": "0DEADC0DE", "from": "3️⃣7️⃣3️⃣5️⃣9️⃣2️⃣9️⃣0️⃣5️⃣4️⃣"},
			},
		}
		testConverter(t, converter)
	}
	{
		b, err := NewBaseConversion(
			Hexadecimal, Binary,
			options.BaseConversion().SetZeroPadding(false),
		)
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b,
			"testCases": []map[string]string{
				{"from": "80", "to": "10000000"},
				{"from": "91", "to": "10010001"},
				{"from": "A2", "to": "10100010"},
				{"from": "B3", "to": "10110011"},
				{"from": "C4", "to": "11000100"},
				{"from": "D5", "to": "11010101"},
				{"from": "E6", "to": "11100110"},
				{"from": "F7", "to": "11110111"},
				{"from": "08", "to": "1000"},
				{"from": "19", "to": "11001"},
				{"from": "2A", "to": "101010"},
				{"from": "3B", "to": "111011"},
				{"from": "4C", "to": "1001100"},
				{"from": "5D", "to": "1011101"},
				{"from": "6E", "to": "1101110"},
				{"from": "7F", "to": "1111111"},
			},
		}
		testConverter(t, converter)
	}
	{
		b, err := NewBaseConversion(
			Hexadecimal, Binary,
			options.BaseConversion().SetZeroPadding(false),
		)
		if err != nil {
			t.Fatal(err)
		}
		var converter = map[string]interface{}{
			"instance": b.Inverse(),
			"testCases": []map[string]string{
				{"to": "80", "from": "10000000"},
				{"to": "91", "from": "10010001"},
				{"to": "A2", "from": "10100010"},
				{"to": "B3", "from": "10110011"},
				{"to": "C4", "from": "11000100"},
				{"to": "D5", "from": "11010101"},
				{"to": "E6", "from": "11100110"},
				{"to": "F7", "from": "11110111"},
				{"to": "8", "from": "00001000"},
				{"to": "19", "from": "00011001"},
				{"to": "2A", "from": "00101010"},
				{"to": "3B", "from": "00111011"},
				{"to": "4C", "from": "01001100"},
				{"to": "5D", "from": "01011101"},
				{"to": "6E", "from": "01101110"},
				{"to": "7F", "from": "01111111"},
			},
		}
		testConverter(t, converter)
	}
}

func testConverter(t *testing.T, converter map[string]interface{}) {
	t.Run("Convert", func(t *testing.T) {
		for _, testCase := range converter["testCases"].([]map[string]string) {
			t.Run(fmt.Sprintf("%v -> %v", testCase["from"], testCase["to"]), func(t *testing.T) {
				actual, err := (converter["instance"]).(*baseConversion).Convert(testCase["from"])
				if err != nil {
					t.Fatal(err)
				}
				expected := testCase["to"]
				if actual != expected {
					t.Fatal()
				}
			})
		}
	})
}


base_conversion_bench_test.go
package baseconv

import (
	"testing"
)

func BenchmarkASCIIConverter(bm *testing.B) {
	b, err := NewBaseConversion("0123456789", Hexadecimal)
	if err != nil {
		bm.Fatal(err)
	}

	s := "33204"

	bm.ResetTimer()
	for i := 0; i < bm.N; i++ { // 1366 ns/op
		_, err := b.Convert(s)
		if err != nil {
			bm.Fatal(err)
		}
	}
}

func BenchmarkNonASCIIConverter(bm *testing.B) {
	b, err := NewBaseConversion("౦123456789", Hexadecimal)
	if err != nil {
		bm.Fatal(err)
	}

	s := "332౦4"

	bm.ResetTimer()
	for i := 0; i < bm.N; i++ { // 4104 ns/op
		_, err := b.Convert(s)
		if err != nil {
			bm.Fatal(err)
		}
	}
}

func BenchmarkIntegerLengthFactorConverter(bm *testing.B) {
	bm.Run("Base64 -> Base8", func(bm *testing.B) {
		b, err := NewBaseConversion(Base64, Base8)
		if err != nil {
			bm.Fatal(err)
		}

		s := "33204"

		bm.ResetTimer()
		for i := 0; i < bm.N; i++ { // 1617 ns/op
			_, err := b.Convert(s)
			if err != nil {
				bm.Fatal(err)
			}
		}
	})

	bm.Run("Base10 -> Base10 (reverted)", func(bm *testing.B) {
		b, err := NewBaseConversion(Base10, "9876543210")
		if err != nil {
			bm.Fatal(err)
		}

		s := "33204"

		bm.ResetTimer()
		for i := 0; i < bm.N; i++ { // 811.8 ns/op
			_, err := b.Convert(s)
			if err != nil {
				bm.Fatal(err)
			}
		}
	})

	bm.Run("Base6 -> Base36", func(bm *testing.B) {
		b, err := NewBaseConversion(Base6, Base36)
		if err != nil {
			bm.Fatal(err)
		}

		s := "33204"

		bm.ResetTimer()
		for i := 0; i < bm.N; i++ { // 1081 ns/op
			_, err := b.Convert(s)
			if err != nil {
				bm.Fatal(err)
			}
		}
	})
}

func BenchmarkNonIntegerLengthFactorConverter(bm *testing.B) {
	bm.Run("Base64 -> Base6", func(bm *testing.B) {
		b, err := NewBaseConversion(Base64, Base6)
		if err != nil {
			bm.Fatal(err)
		}

		s := "33204"

		bm.ResetTimer()
		for i := 0; i < bm.N; i++ { // 2564 ns/op
			_, err := b.Convert(s)
			if err != nil {
				bm.Fatal(err)
			}
		}
	})

	bm.Run("Base64 -> Base10", func(bm *testing.B) {
		b, err := NewBaseConversion(Base64, Base10)
		if err != nil {
			bm.Fatal(err)
		}

		s := "33204"

		bm.ResetTimer()
		for i := 0; i < bm.N; i++ { // 2145 ns/op
			_, err := b.Convert(s)
			if err != nil {
				bm.Fatal(err)
			}
		}
	})

	bm.Run("Base5 -> Base36", func(bm *testing.B) {
		b, err := NewBaseConversion("01234", Base36)
		if err != nil {
			bm.Fatal(err)
		}

		s := "33204"

		bm.ResetTimer()
		for i := 0; i < bm.N; i++ { // 1239 ns/op
			_, err := b.Convert(s)
			if err != nil {
				bm.Fatal(err)
			}
		}
	})

	bm.Run("Base7 -> Base36", func(bm *testing.B) {
		b, err := NewBaseConversion("0123456", Base36)
		if err != nil {
			bm.Fatal(err)
		}

		s := "33204"

		bm.ResetTimer()
		for i := 0; i < bm.N; i++ { // 1242 ns/op
			_, err := b.Convert(s)
			if err != nil {
				bm.Fatal(err)
			}
		}
	})
}


example/main.go
package main

import (
	"fmt"

	"go.dkinom.dev/baseconv"
	"go.dkinom.dev/baseconv/options"
)

func main() {
	decimalAndBinary()
	decimalAndDecimalEmoji()
	decimalEmojiAndHexadecimal()
}

func decimalAndBinary() {
	// Decimal alphabet - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
	// Binary alphabet - 0, 1
	decimalToBinaryConverter, err := baseconv.NewBaseConversion(baseconv.Base10, baseconv.Base2)
	if err != nil {
		panic(err)
	}

	binary, err := decimalToBinaryConverter.Convert("3888")
	if err != nil {
		panic(err)
	}
	fmt.Printf("decimal number 3888 is represented as %v in binary\n", binary) // 111100110000

	decimal, err := decimalToBinaryConverter.Inverse().Convert("111110011111")
	if err != nil {
		panic(err)
	}
	fmt.Printf("binary number 111110011111 is represented as %v in decimal\n", decimal) // 3999
}

func decimalAndDecimalEmoji() {
	// Decimal alphabet - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
	// Decimal emoji alphabet - 0️⃣, 1️⃣, 2️⃣, 3️⃣, 4️⃣, 5️⃣, 6️⃣, 7️⃣, 8️⃣, 9️⃣
	decimalToDecimalEmojiConverter, err := baseconv.NewBaseConversion(baseconv.Base10, "0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣")
	if err != nil {
		panic(err)
	}

	decimalEmoji, err := decimalToDecimalEmojiConverter.Convert("38888")
	if err != nil {
		panic(err)
	}
	fmt.Printf("decimal number 38888 is represented as %v in decimal emoji\n", decimalEmoji) // 3️⃣8️⃣8️⃣8️⃣8️⃣

	decimal, err := decimalToDecimalEmojiConverter.Inverse().Convert("3️⃣9️⃣9️⃣9️⃣9️⃣")
	if err != nil {
		panic(err)
	}
	fmt.Printf("decimal emoji number 3️⃣9️⃣9️⃣9️⃣9️⃣ is represented as %v in decimal\n", decimal) // 39999
}

func decimalEmojiAndHexadecimal() {
	// Decimal emoji alphabet - 0️⃣, 1️⃣, 2️⃣, 3️⃣, 4️⃣, 5️⃣, 6️⃣, 7️⃣, 8️⃣, 9️⃣
	// Hexadecimal alphabet - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, A, B, C, D, E, F
	decimalEmojiToHexadecimalConverter, err :=
		baseconv.NewBaseConversion(
			"0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣", baseconv.Base16,
			options.BaseConversion().
				SetZeroPadding(true),
		)
	if err != nil {
		panic(err)
	}

	hexadecimal, err := decimalEmojiToHexadecimalConverter.Convert("5️⃣1️⃣9️⃣6️⃣6️⃣")
	if err != nil {
		panic(err)
	}
	fmt.Printf("decimal emoji number 5️⃣1️⃣9️⃣6️⃣6️⃣ is represented as %v in hexadecimal\n", hexadecimal) // 0CAFE

	decimalEmoji, err := decimalEmojiToHexadecimalConverter.Inverse().Convert("DEADC0DE")
	if err != nil {
		panic(err)
	}
	fmt.Printf(
		"hexadecimal number DEADC0DE is represented as %v in decimal emoji\n", decimalEmoji,
	) // 3️⃣7️⃣3️⃣5️⃣9️⃣2️⃣9️⃣0️⃣5️⃣4️⃣
}


base_conversion.go
package baseconv

import (
	"fmt"
	"math"
	"strings"

	"go.dkinom.dev/baseconv/internal/characters"
	"go.dkinom.dev/baseconv/options"
)

// Provides ability to convert numbers between different
// positional numeral systems
type baseConversion struct {
	fromAlphabet alphabet
	toAlphabet   alphabet

	lengthFactor    float64
	intLengthFactor int

	zeroPadding bool
}

// Initializes baseConversion struct with the given from and to alphabets
func NewBaseConversion(from string, to string, opts ...*options.BaseConversionOptions) (*baseConversion, error) {
	fromAlphabet, err := NewAlphabet(from)
	if err != nil {
		return nil, err
	}
	toAlphabet, err := NewAlphabet(to)
	if err != nil {
		return nil, err
	}

	b := NewBaseConversionAlphabet(*fromAlphabet, *toAlphabet, opts...)

	return b, nil
}

// Initializes baseConversion struct with the given from and to alphabets
func NewBaseConversionAlphabet(from alphabet, to alphabet, opts ...*options.BaseConversionOptions) *baseConversion {
	b := &baseConversion{fromAlphabet: from, toAlphabet: to}

	b.lengthFactor = math.Log(float64(from.Radix())) / math.Log(float64(to.Radix()))

	if b.lengthFactor == math.Trunc(b.lengthFactor) {
		b.intLengthFactor = int(b.lengthFactor)
	} else {
		invLengthFactor := (1 / b.lengthFactor)
		if invLengthFactor == math.Trunc(invLengthFactor) {
			b.intLengthFactor = -int(invLengthFactor)
		}
	}

	opt := options.MergeBaseConversionOptions(opts...)

	b.configure(opt)

	return b
}

func (b *baseConversion) configure(opt *options.BaseConversionOptions) {
	// Zero padding
	b.zeroPadding = false
	if opt.ZeroPadding != nil {
		b.zeroPadding = *opt.ZeroPadding
	}
}

func (b *baseConversion) options() *options.BaseConversionOptions {
	o := options.BaseConversion()

	// Zero padding
	o.SetZeroPadding(b.zeroPadding)

	return o
}

// Returns numeral representation of s in toAlphabet
func (b *baseConversion) Convert(s string) (r string, err error) {
	if !b.fromAlphabet.IsValid(s) {
		err = fmt.Errorf("'%v' is invalid", s)
		return
	}

	r = b.convertIntegralPart(s)
	return
}

func (b *baseConversion) convertIntegralPart(ip string) string {
	fromBase := b.fromAlphabet.Radix()
	toBase := b.toAlphabet.Radix()

	var changeBase func(_values []int) (string, int)
	changeBase = func(_values []int) (string, int) {
		values := []int{}
		for i, v := range _values {
			if v != 0 {
				values = append(values, _values[i:]...)
				break
			}
		}

		if len(values) == 0 {
			return "", 0
		}

		if b.intLengthFactor == 1 || b.intLengthFactor == -1 {
			sb := strings.Builder{}
			for _, value := range values {
				sb.WriteString(b.toAlphabet.chars[value])
			}

			return sb.String(), len(values)
		} else if b.intLengthFactor == 0 ||
			len(values) == 1 ||
			(b.intLengthFactor < 0 && b.intLengthFactor+len(values) <= 0) { // Horner's method
			remainder := 0
			quotients := []int{}
			for _, value := range values {
				remainder = (remainder * fromBase) + value

				quotients = append(quotients, remainder/toBase)
				remainder = remainder % toBase
			}

			r, rLen := changeBase(quotients)

			return r + b.toAlphabet.chars[remainder], rLen + 1
		} else if b.intLengthFactor > 0 {
			sb, rLen := strings.Builder{}, 0
			for i := 0; i < len(values); i++ {
				_r, _rLen := changeBase(values[i : i+1])
				if i > 0 {
					sb.WriteString(strings.Repeat(b.toAlphabet.zerochar, b.intLengthFactor-_rLen))
					sb.WriteString(_r)
					rLen += b.intLengthFactor
				} else {
					sb.WriteString(_r)
					rLen += _rLen
				}
			}

			return sb.String(), rLen
		} else { // b.intLengthFactor < 0
			sb, rLen := strings.Builder{}, 0
			i := len(values) % (-b.intLengthFactor)
			if i > 0 {
				i += b.intLengthFactor
			}
			for ; i < len(values); i -= b.intLengthFactor {
				var _values []int
				if i > 0 {
					_values = values[i : i-b.intLengthFactor]
					_r, _rLen := changeBase(_values)
					sb.WriteString(strings.Repeat(b.toAlphabet.zerochar, 1-_rLen))
					sb.WriteString(_r)
					rLen += 1
				} else {
					_values = values[:i-b.intLengthFactor]
					_r, _rLen := changeBase(_values)
					sb.WriteString(_r)
					rLen += _rLen
				}
			}

			return sb.String(), rLen
		}
	}

	values := []int{}
	c := characters.NewCharacters(ip)
	for c.Next() {
		char := c.Str()

		values = append(values, b.fromAlphabet.charset[char])
	}
	ipLen := len(values)

	r, rLen := changeBase(values)

	if b.zeroPadding {
		currentLength := rLen
		wantedLength := int(math.Ceil(float64(ipLen) * b.lengthFactor))

		if currentLength < wantedLength {
			var sb strings.Builder
			sb.WriteString(strings.Repeat(b.toAlphabet.zerochar, wantedLength-currentLength))
			sb.WriteString(r)

			r = sb.String()
		}
	} else if rLen == 0 {
		r = b.toAlphabet.zerochar
	}

	return r
}

// Returns a baseConversion which converts numerals
// from toAlphabet to fromAlphabet
func (b *baseConversion) Inverse() *baseConversion {
	return NewBaseConversionAlphabet(b.toAlphabet, b.fromAlphabet, b.options())
}


alphabet.go
package baseconv

import (
	"fmt"
	"strings"

	"go.dkinom.dev/baseconv/internal/characters"
)

type alphabet struct {
	chars    []string
	charset  map[string]int
	zerochar string
}

func NewAlphabet(str string) (*alphabet, error) {
	var chars []string

	charset := make(map[string]int)

	c := characters.NewCharacters(str)
	for c.Next() {
		char := c.Str()

		chars = append(chars, char)

		charset[char] = len(chars) - 1
	}

	if len(chars) != len(charset) {
		return nil, fmt.Errorf("Must not have duplicate characters in alphabet")
	}

	a := &alphabet{
		chars:    chars,
		charset:  charset,
		zerochar: chars[0],
	}

	return a, nil
}

func (a *alphabet) String() string {
	return strings.Join(a.chars, ",")
}

// Radix of this alphabet
func (a *alphabet) Radix() int {
	return len(a.chars)
}

// Checks if s is a valid numeral representation
func (a *alphabet) IsValid(s string) bool {
	var previousChar *string
	c := characters.NewCharacters(s)
	for c.Next() {
		char := c.Str()

		if _, present := a.charset[char]; !present {
			return false
		}

		previousChar = &char
	}
	return previousChar != nil
}

// Binary numeral system
const Binary = "01"

// Same as Binary numeral system
const Base2 = Binary

// Ternary numeral system
const Ternary = "012"

// Same as Ternary numeral system
const Base3 = Ternary

// Quaternary numeral system
const Quaternary = "0123"

// Same as Quaternary numeral system
const Base4 = Quaternary

// Quinary numeral system
const Quinary = "01234"

// Same as Quinary numeral system
const Base5 = Quinary

// Senary numeral system
const Senary = "012345"

// Same as Senary numeral system
const Base6 = Senary

// Octal numeral system
const Octal = "01234567"

// Same as Octal numeral system
const Base8 = Octal

// Decimal numeral system
const Decimal = "0123456789"

// Same as Decimal numeral system
const Base10 = Decimal

// Duodecimal numeral system
const Duodecimal = "0123456789AB"

// Same as Duodecimal numeral system
const Base12 = Duodecimal

// Hexadecimal numeral system
const Hexadecimal = "0123456789ABCDEF"

// Same as Hexadecimal numeral system
const Base16 = Hexadecimal

// Base32 numeral system
const Base32 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"

// Base32 (Extended Hex) numeral system
const Base32hex = "0123456789ABCDEFGHIJKLMNOPQRSTUV"

// Base36 numeral system
const Base36 = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"

// Base58 numeral system
const Base58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

// Base62 numeral system
const Base62 = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

// Base64 numeral system
const Base64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"

// Base64 (URL and Filename safe) numeral system
const Base64url = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
